pmd-java/src/main/java/net/sourceforge/pmd/lang/java/JavaLanguageModule.java

```diff
@@ -26,7 +26,8 @@ public JavaLanguageModule() {
         addVersion("9", new JavaLanguageHandler(9), false);
         addVersion("10", new JavaLanguageHandler(10), false);
         addVersion("11", new JavaLanguageHandler(11), false);
-        addVersion("12", new JavaLanguageHandler(12), true);
+        addVersion("12", new JavaLanguageHandler(12), false);
```

Comment:
Hm... to quote https://openjdk.java.net/jeps/12:

> A preview language or VM feature is a new feature of the Java SE Platform that is fully specified, fully implemented, and yet impermanent.

There is no guarantee, that the preview feature will stay. It's "use at your own risk" if you want to experiment with the latest and give feedback.

I think, it is important, that PMD doesn't fail to parse Java, if it sees such code. Otherwise, you wouldn't be able to use PMD...

Writing rules for preview features should be delayed - since preview features might be removed with the java version, the rule might become irrelevant very soon.

> instability is to be expected

I consider this as given, if one uses preview features, that he understands, that this is not stable... it's preview. In order to use the preview features, you need to enable it at compile time and again at runtime. So, you for sure do not accidentally use preview features, it's a conscious decision.

> we should continue supporting each preview version until at least the next major release.

Maybe. For this specific case (break-with-return), we could support it. But that might not always be an option.

From the release plan of openjdk, Java12 is not a LTS release (at least, I don't think so). This means, as soon as Java13 is released (2019/09/17), Java12 is unsupported. In order to compile source code with the break-with-return statement, you have to use Java12 - Java13 doesn't support this "old" version of Java anymore... That's the risk of using preview features.

I hope, we can release PMD 6.18.0 just before Java13 is released. We could say, that we still parse with PMD 6.18.0 the Java12 break-with-return statement, but drop it with the next PMD version 6.19.0?
I'd like to avoid maintaining this longer than actually needed, since the feature makes only sense with Java12 and no other version of Java... which will be gone, once Java13 is there.

Reply:
It seems we can simply revert 2ba1422 for this release...

Reply:
I tend to agree with @adangel on this one… I'd not like to indefinitely keep supporting language features that are specific to a single release and need to be enabled to be used.

I think we, as PMD, should try to support preview features mainly for 2 reasons:
1. Be able to give feedback to the Java community
2. Be ready for the time those features mature and graduate out of preview, knowing some may falter along the way

I think we can commit to only supporting preview features for the latest Java release (if any at all) and that should not surprise our users, although we can probably make a better job at advertising that promise.

Reply:
I've reverted the mentioned commit and updated the release notes.
I think, this PR is now ready to merge.

We'll keep supporting java12 preview for PMD 6.18.0, but drop support for it with the next version PMD 6.19.0.

Btw - I finally found one reference, when to expect the next LTS java release:

> In addition, every three years one feature release will be designated as the Long Term 
> Supported (LTS) release. 

(https://adoptopenjdk.net/support.html#roadmap)

Seems like in 2021, there will be a next java LTS release.

Reply:
> we can probably make a better job at advertising that promise.

Introducing a new language version for preview features would be the best kind of advertisement. It wouldn't be the default version for Java, and would allow users to scope rules specific to preview versions just to that preview version. For now we have a "13" language version which implicitly enables preview features, and the code of the rules has no way to know if preview is enabled or not (PMD_JAVA_OPTS is a VM option, not actually given to pmd anywhere). Switch expressions are still a preview feature of jdk 12, which we're not disabling because it carried on to jdk 13 - yet it's still not a java 12 feature. It all seems very murky to not distinguish language levels with a language version here.

Reply:
I understand your point. How about this suggestion:

* We set the default language version back to 11. There were no language changes in 12 and 13, except for the preview languages features. So we can safely go back from 12 to 11 without any incompatibilities.
* We introduce in JavaLanguageModule a version "12.preview" and "13.preview" (that will introduce changes down the chain in JavaLanguageHandler as well to use more than a simple integer to describe the version).
* In order to use the preview features, this means: compile with preview, run with preview and select the preview version for PMD as well.

Instead of 12.preview we could also use something like 12.65535 - I think, that's technically the version of the class file format, when the preview feature is enabled.

I can have a look tonight/tomorrow in this....

@oowekyala: wdyt?

Reply:
> We set the default language version back to 11. There were no language changes in 12 and 13, except for the preview languages features. So we can safely go back from 12 to 11 without any incompatibilities.

My understanding of language versions is that they don't only mean language features, but also for Java, the JDK API version. IIRC we have some rules/planned rules that are disabled if the API version is less than the relevant `@since` (eg #1034 only applies on JDK 8+). So maybe the default should change to 13? Or should it be just the latest LTS release?

I fully agree with your other points (imho "12.preview" seems less cryptic btw)

Reply:
> for Java, the JDK API version

oh yes, I didn't think of that. Yes, this means, we need to have version 12 and 13. While there are no grammar changes (putting preview aside), there might have been API changes.

The question is then - which should be the default? We usually preselected always the latest version, but I think mostly due to the fact, that "it doesn't hurt", since Java is always backwards compatible (except for cases, when assert or enum were made to keywords). I still tend to just preselect the latest, even though it might not be a LTS release -> 13

Reply:
Yeah I think taking the latest non-preview release as default is ok, and is what we've been doing for a while